### ISSUE TITLE
[Bug] Binding with arrays sometimes doesn't work as expected

### ISSUE THREAD

**Issue by jaeko44** [aurelia/binding#472]
**I'm submitting a bug report**

* **Library Version:**
major

**Please tell us about your environment:**
* **Operating System:**
Windows [10]

* **Node Version:**
6.2.0

* **NPM Version:**
3.8.9

* **JSPM OR Webpack AND Version**
au-cli - 0.17.0

* **Browser:**
Chrome XX

* **Language:**
ESNext

**Current behavior:**
https://gist.run/?id=680d7ef5a0f04432d1753763469d5145

**Expected/desired behavior:**
Update to blue when the text is clicked (changeColor() function)

* **What is the motivation / use case for changing the behavior?**
AFAIK the functionality of 'bind' should be one that checks for changes and automatically. If that's not the purpose of 'bind' then please point me to the correct dierction.

**Comment by jaeko44** [comment]
I've encountered this bug with value.bind="tempMessage[chat.id]"

If i have an array with multiple 'tempMessages' (Using this as the UI is being built with a loop and each message is it's 'own')

Now, when I actually enter data it is being written to 'tempMessage[chat.id]' just fine, once I do a submit and reset tempMessage[chat.id] = '';

The input doesn't clear to '', whatever was written is still there. If I leave the input and click submit again, then nothing is being sent. 

I've taken a video to show this clearly. 

https://drive.google.com/file/d/0B_5RWlbn0wSOTDh5WUx4VnAxTDA/view

**Comment by EisenbergEffect** [comment]
@jdanyow Is this documented in the binding docs? If not we should add something on limitations and workarounds.

**Comment by jaeko44** [comment]
@EisenbergEffect 

I ran into a 3rd bug/limitation with .bind, this time it is with the repeat.for method in views.

I have the 3 chats being repeated as above, now I've added a method that removes the last chat from the array and adds a new chat in the beginning. 

Now, while testing the function manually (from the same class) it was working nicely and smoothly. 

However when implementing it from the navbar (to open chats accordingly) - the function runs as expected, however the view doesn't detect the array being modified and it update in the view..
